content-scripts/cs.js

```diff
@@ -366,33 +383,79 @@ async function onInfoAvailable({ globalState: globalStateMsg, addonsWithUserscri
-        if (!dynamicDisable) return;
-        // Addon was reenabled
-        _page_.fireEvent({ name: "reenabled", addonId, target: "self" });
+      if (partial) {
+        // Partial: part of userstyle was (re-)enabled.
+        // No need to deal with userscripts here.
+        const addonsWithUserstylesEntry = addonsWithUserscripts.find((entry) => entry.addonId === addonId);
+        if (addonsWithUserstylesEntry) {
+          addonsWithUserstylesEntry.styles = userstyles;
+        } else {
+          addonsWithUserstyles.push({ styles: userstyles, cssVariables, addonId, injectAsStyleElt, index });
+        }
       } else {
-        if (!dynamicEnable) return;
-        // Addon was not injected in page yet
-        _page_.runAddonUserscripts({ addonId, scripts, enabledLate: true });
-      }
+        // Non-partial: the whole addon was (re-)enabled.
+        if (everLoadedUserscriptAddons.has(addonId)) {
+          if (!dynamicDisable) return;
+          // Addon was reenabled
+          _page_.fireEvent({ name: "reenabled", addonId, target: "self" });
+        } else {
+          if (!dynamicEnable) return;
+          // Addon was not injected in page yet
+          _page_.runAddonUserscripts({ addonId, scripts, enabledLate: true });
+          everLoadedUserscriptAddons.add(addonId);
+        }
 
-      addonsWithUserscripts.push({ addonId, scripts });
-      addonsWithUserstyles.push({ styles: userstyles, cssVariables, addonId, injectAsStyleElt, index });
+        addonsWithUserscripts.push({ addonId, scripts });
+        addonsWithUserstyles.push({ styles: userstyles, cssVariables, addonId, injectAsStyleElt, index });
+      }
       setCssVariables(globalState.addonSettings, addonsWithUserstyles);
-      everLoadedAddons.push({ addonId, scripts });
     } else if (request.dynamicAddonDisable) {
-      const { addonId } = request.dynamicAddonDisable;
-      disabledDynamicAddons.push(addonId);
+      // Note: partialDynamicDisabledStyles includes ones that are disabled currently, too!
+      const { addonId, partialDynamicDisabledStyles } = request.dynamicAddonDisable;
+      // This may run twice if the style-only addon was first "partially"
+      // (but in fact entirely) disabled, and it was then toggled off.
+      // Early return in this situation.
+      if (disabledDynamicAddons.has(addonId)) return;
+      const scriptIndex = addonsWithUserscripts.findIndex((a) => a.addonId === addonId);
+      const styleIndex = addonsWithUserstyles.findIndex((a) => a.addonId === addonId);
+      if (partialDynamicDisabledStyles) {
+        // Userstyles are partially disabled.
+        // This should not result in other parts being disabled,
+        // unless that means no scripts/styles are running on this page.
+        removeAddonStylesPartial(addonId, partialDynamicDisabledStyles);
+        if (styleIndex > -1) {
+          // This should exist... right? Safeguarding anyway
```

Comment:
I'm reaching this if condition.

Reply:
See repro in issue comment (not code comment) below to see how to reach this if case.